packages/flutter_tools/lib/src/reporting/unified_analytics.dart

```diff
@@ -0,0 +1,52 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:unified_analytics/unified_analytics.dart';
+
+import '../version.dart';
+
+/// This function is called from within the context runner to perform
+/// checks that are necessary for determining if a no-op version of
+/// [Analytics] gets returned.
+///
+/// When [enableAsserts] is set to `true`, various assert statements
+/// will be enabled to ensure usage of this class is within GA4 limitations.
+///
+/// For testing purposes, pass in a [FakeAnalytics] instance initialized with
+/// an in-memory [FileSystem] to prevent writing to disk.
+Analytics getAnalytics({
+  required bool runningOnBot,
+  required FlutterVersion flutterVersion,
+  required Map<String, String> environment,
+  bool enableAsserts = false,
+  FakeAnalytics? analyticsOverride,
+}) {
+  final String version = flutterVersion.getVersionString(redactUnknownBranches: true);
+  final bool suppressEnvFlag = environment['FLUTTER_SUPPRESS_ANALYTICS']?.toLowerCase() == 'true';
+
+  if (// Ignore local user branches.
+      version.startsWith('[user-branch]') ||
+      // Many CI systems don't do a full git checkout.
+      version.endsWith('/unknown') ||
+      // Ignore bots.
+      runningOnBot ||
+      // Ignore when suppressed by FLUTTER_SUPPRESS_ANALYTICS.
+      suppressEnvFlag) {
+    return NoOpAnalytics();
+  }
+
+  // Providing an override of the [Analytics] instance is preferred when
+  // running tests for this function to prevent writing to the filesystem
+  if (analyticsOverride != null) {
```

Comment:
This concerns me if this is the only safeguard that we have that we're not accessing the real filesystem and making real network requests. Let's discuss offline.

Reply:
I think it will be sufficient if we ensure the `testUsingContext()` function overrides this value in the context.